### PR TITLE
Complete draft for property based testing (Dec 9) - invitation to copy-edit

### PR DESCRIPTION
**Pull request by chewxy**
There's about 1500 words more to go.

### CODE REVIEW

File: content/advent-2017/property-based-testing.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+func foo() {
+	returnsError() // bad! Don't do this! Always check for and handle errors!
+}
+```
+
+But let's say you are a diligent programmer, and you checked all the errors, other forms of errors may occur too: panics from slice indices being out of bounds, etc. 
+
+So an idea emerges: why don't we test on all possible values? Or at least as many values as we can? 
+
+Enter the notion of [fuzz testing](https://en.wikipedia.org/wiki/Fuzzing): we feed the program random inputs, and then watch for it to fail. Fuzz testing often leads one to find subtle bugs you don't expect. Here's one that Chris Marshall discovered on my [skiprope](https://github.com/chewxy/skiprope) library. I have to this day, no idea what the correct fix is. Solutions welcome
```
Comment (dgryski) [inline content/advent-2017/property-based-testing.md]: Missing link to actual `skiprope` issue?

File: content/advent-2017/property-based-testing.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+		t.Error(err)
+	}
+}
+```
+
+Here we see that `comm` was defined rather simply: `if Add(a, b) != Add(b, a)` then we say that the `Add` function does not fulfil the commutativity property.
+
+What the `quick.Check` function then does is it generates values based on the input arguments of the testing function. In this case, `comm` is a function that takes two `int`. The package understands that it needs to generate two `int` to be fed into the function. 
+
+The package works for non-primitive types too. Further extensions to functionality can be had by types that implement the `quick.Generator` interface.
```
Comment (dgryski) [inline content/advent-2017/property-based-testing.md]: Give small example of something with the `Generator` interface.  Maybe integers that are only even or something?  I've also used it for to generate slices of integers with smaller abolute values to make them easier to read.

File: content/advent-2017/property-based-testing.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+For example, let's say you invented a new sort algorithm. What properties that all sort algorithms have to hold in order to be useful? Well, for one, the container size should not change. Imagine a sorting algorithm that sorts a list and the result has fewer or more elements than when it started. Surely something fishy is happening there.
+
+As another example, recently I wrote a parser. What are the properties of a parser? One thing I tested was that I was able to retrieve the input from the parser - that is to say, I'm able to make the input and output the same. That test found a bunch of bugs in the parsing logic that I have yet to fix.
+
+## Property Based Testing You Probably Already Do ##
+
+One of the interesting things about property based testing is that you probably already do some of it. Take the `gob` package for example. The example in the gob package is a form of property based testing - granted it's not tested a large variety of inputs, but the example is an excellent starting point of a property based testing methodology.
```
Comment (dgryski) [inline content/advent-2017/property-based-testing.md]: Link to or describe what `gob` does?

File: content/advent-2017/property-based-testing.md (hunk 1 above)
Comment (chewxy) [reply content/advent-2017/property-based-testing.md]: thanks for the spot. Will be in next commit

File: content/advent-2017/property-based-testing.md (hunk 2 above)
Comment (chewxy) [reply content/advent-2017/property-based-testing.md]: ditto to the above

Comment (chewxy) [reply content/advent-2017/property-based-testing.md]: Fixed

**Comment by bketelsen** [comment]
does this need any more review/editing or is it ready to publish tomorrow?

**Comment by chewxy** [comment]
Well if no one has anything to add, then it's good to go

**Pull request merged**